tket/src/Circuit/Multiplexor.cpp

```diff
@@ -0,0 +1,668 @@
+  for (auto it = op_map.begin(); it != op_map.end(); it++) {
+    SymSet op_symbols = it->second->free_symbols();
+    all_symbols.insert(op_symbols.begin(), op_symbols.end());
+  }
+  return all_symbols;
+}
+
+static ctrl_op_map_t op_map_dagger(const ctrl_op_map_t &op_map) {
+  ctrl_op_map_t new_op_map;
+  for (auto it = op_map.begin(); it != op_map.end(); it++) {
+    new_op_map.insert({it->first, it->second->dagger()});
+  }
+  return new_op_map;
+}
+
+static ctrl_op_map_t op_map_transpose(const ctrl_op_map_t &op_map) {
+  ctrl_op_map_t new_op_map;
+  for (auto it = op_map.begin(); it != op_map.end(); it++) {
+    new_op_map.insert({it->first, it->second->transpose()});
+  }
+  return new_op_map;
+}
+
+UniformQControlBox::UniformQControlBox(const ctrl_op_map_t &op_map)
+    : Box(OpType::UniformQControlBox), op_map_(op_map) {
+  auto it = op_map.begin();
+  if (it == op_map.end()) {
+    throw std::invalid_argument("No Ops provided.");
+  }
+  n_controls_ = (unsigned)it->first.size();
+  n_targets_ = it->second->n_qubits();
+  op_map_validate(op_map);
+}
+
+UniformQControlBox::UniformQControlBox(const UniformQControlBox &other)
+    : Box(other),
+      n_controls_(other.n_controls_),
+      n_targets_(other.n_targets_),
+      op_map_(other.op_map_) {}
+
+Op_ptr UniformQControlBox::symbol_substitution(
+    const SymEngine::map_basic_basic &sub_map) const {
+  ctrl_op_map_t new_op_map = op_map_symbol_sub(sub_map, op_map_);
+  return std::make_shared<UniformQControlBox>(new_op_map);
+}
+
+SymSet UniformQControlBox::free_symbols() const {
+  return op_map_free_symbols(op_map_);
+}
+
+Op_ptr UniformQControlBox::dagger() const {
+  return std::make_shared<UniformQControlBox>(op_map_dagger(op_map_));
+}
+
+Op_ptr UniformQControlBox::transpose() const {
+  return std::make_shared<UniformQControlBox>(op_map_transpose(op_map_));
+}
+
+op_signature_t UniformQControlBox::get_signature() const {
+  op_signature_t qubits(n_controls_ + n_targets_, EdgeType::Quantum);
+  return qubits;
+}
+
+nlohmann::json UniformQControlBox::to_json(const Op_ptr &op) {
+  const auto &box = static_cast<const UniformQControlBox &>(*op);
+  nlohmann::json j = core_box_json(box);
+  j["op_map"] = box.get_op_map();
+  return j;
+}
+
+Op_ptr UniformQControlBox::from_json(const nlohmann::json &j) {
+  UniformQControlBox box =
+      UniformQControlBox(j.at("op_map").get<ctrl_op_map_t>());
+  return set_box_id(
+      box,
+      boost::lexical_cast<boost::uuids::uuid>(j.at("id").get<std::string>()));
+}
+
+void UniformQControlBox::generate_circuit() const {
+  circ_ = std::make_shared<Circuit>(
+      multiplexor_sequential_decomp(op_map_, n_controls_, n_targets_));
+}
+
+UniformQControlRotationBox::UniformQControlRotationBox(
+    const ctrl_op_map_t &op_map)
+    : Box(OpType::UniformQControlRotationBox), op_map_(op_map) {
+  auto it = op_map.begin();
+  if (it == op_map.end()) {
+    throw std::invalid_argument("No Ops provided.");
+  }
+  n_targets_ = 1;
```

Comment:
Does this need to be an attribute or can the "1' just be used as a hardcoded value elsewhere?
tbh n_targets_ is probably more helpful

Reply:
They should be removed. At first I intended to make the U2, and the rotation multiplexors children class of the general multiplexor, and I forgot to removing them after changing my mind.